LinuxCommunicator/HostsFile/HostsFileManager.cs

```diff
@@ -0,0 +1,207 @@
+        public void ReloadTimerEvent(object param)
+        {
+            try
+            {
+                FileInfo fileInfo = new FileInfo(this.filepath);
+                if(!fileInfo.Exists)
+                {
+                    LinuxCommunicator.Instance.Tracer.TraceInfo("[HostsFileManager] The hosts file doesn't exists: {0}", this.filepath);
+                    return;
+                }
+
+                if(fileInfo.LastWriteTimeUtc <= this.lastModified)
+                {
+                    LinuxCommunicator.Instance.Tracer.TraceInfo("[HostsFileManager] The hosts file isn't changed since last load");
+                    return;
+                }
+
+                bool manageFile = false;
+                List<HostEntry> newEntries = new List<HostEntry>();
+                foreach (var line in File.ReadAllLines(this.filepath))
+                {
+                    Match commentMatch = HostsFileManager.Comment.Match(line);
+                    if (commentMatch.Success)
+                    {
+                        // throw away comments for the moment.
+                    }
+
+                    Match commentParameterMatch = HostsFileManager.CommentParameter.Match(line);
+                    if (commentParameterMatch.Success && string.Equals(commentParameterMatch.Groups["parameter"].ToString(), "ManageFile", StringComparison.OrdinalIgnoreCase))
+                    {
+                        if (string.Equals(commentParameterMatch.Groups["value"].ToString(), "true", StringComparison.OrdinalIgnoreCase))
+                        {
+                            manageFile = true;
+                        }
+                    }
+
+                    Match ipEntryMatch = HostsFileManager.IpEntry.Match(line);
+                    if (ipEntryMatch.Success && manageFile)
+                    {
+                        string ip = ipEntryMatch.Groups["ip"].ToString();
+                        string name = ipEntryMatch.Groups["dnsName"].ToString();
+                        string comment = ipEntryMatch.Groups["comment"].ToString();
+
+                        if (comment.Equals(HostsFileManager.ManagedEntryKey, StringComparison.OrdinalIgnoreCase))
+                        {
+                            try
+                            {
+                                newEntries.Add(new HostEntry(name, ip));
+                            }
+                            catch (ArgumentException ex)
+                            {
+                                LinuxCommunicator.Instance.Tracer.TraceInfo("[HostsFileManager] Skip invalid host entry name={0}, ip={1}: {2}", name, ip, ex.Message);
+                            }
+                        }
+                    }
+                }
+
+                if (manageFile)
+                {
+                    if (newEntries.Count != this.managedEntries.Count || newEntries.Except(this.managedEntries).Any() || this.managedEntries.Except(newEntries).Any())
```

Comment:
This is not a good condition, and also the third condition is redundant.
You can use new HashSet(newEntries).SetEqual(new ....)
or if you accept N*log(N), you can use newEntries.OrderBy(..).SequenceEqual(...OrderBy())